src/core/algorithms/md/md_verifier/validation/validation.cpp

```diff
@@ -19,48 +51,43 @@ ColumnInfoView MDValidationCalculator::GetColumnInfo(
     auto const& [left_clusters, right_clusters, similarity_matrix] =
             GetColumnInfo(column_match_info);
-    for (model::Index left_cluster_index : hymd::utility::IndexRange(similarity_matrix.size())) {
-        for (auto const [right_cluster_index, similarity_index] :
-             similarity_matrix[left_cluster_index]) {
-            model::md::Similarity similarity =
-                    column_match_info.similarity_info.classifier_values[similarity_index];
-            if (similarity >= decision_boundary) {
-                lam(left_clusters[left_cluster_index], right_clusters[right_cluster_index]);
+    for (hymd::ValueIdentifier left_value_id : hymd::utility::IndexRange(left_clusters.size())) {
+        for (hymd::ValueIdentifier right_value_id :
+             hymd::utility::IndexRange(right_clusters.size())) {
+            model::md::Similarity similarity = 0.0;
+            hymd::indexes::SimilarityMatrixRow const& similarity_matrix_row =
+                    similarity_matrix[left_value_id];
+            if (auto it = similarity_matrix_row.find(right_value_id);
+                it != similarity_matrix_row.end()) {
+                model::Index ccv_id = it->second;
+                similarity = column_match_info.similarity_info.classifier_values[ccv_id];
+            }
+            if (similarity < decision_boundary) {
+                for_each_unmatched(left_clusters[left_value_id], right_clusters[right_value_id],
+                                   similarity);
             }
         }
     }
 }
 
-void MDValidationCalculator::InitRecords(hymd::ColumnMatchInfo const& column_match_info,
-                                         model::md::DecisionBoundary decision_boundary) {
-    auto insert_clusters = [this](hymd::indexes::PliCluster left_cluster,
-                                  hymd::indexes::PliCluster right_cluster) {
-        violating_records_.InsertClusters(left_cluster, right_cluster);
-    };
-    ProcessSimilarityMatrix(column_match_info, decision_boundary, insert_clusters);
-}
-
-void MDValidationCalculator::UpdateRecordsWithLhs(hymd::ColumnMatchInfo const& column_match_info,
-                                                  model::md::DecisionBoundary decision_boundary) {
-    IntersectionBuilder intersection_builder(violating_records_);
-
-    auto add_intersection = [&intersection_builder](hymd::indexes::PliCluster left_cluster,
-                                                    hymd::indexes::PliCluster right_cluster) {
-        intersection_builder.AddIntersection(left_cluster, right_cluster);
+void MDValidationCalculator::RemoveNonMatchedLhsPairs(
+        hymd::ColumnMatchInfo const& column_match_info,
+        model::md::DecisionBoundary decision_boundary) {
+    auto remove_unmatched = [this](hymd::indexes::PliCluster left_cluster,
+                                   hymd::indexes::PliCluster right_cluster,
+                                   [[maybe_unused]] model::md::Similarity similarity) {
```

Comment:
`[[maybe_unused]]` is for stuff like variables disappearing in macros.
```suggestion
                                   model::md::Similarity) {
```